### ISSUE TITLE
Use LZMA compression for transmission of binary packages

### ISSUE THREAD

**Issue by annulen** [conan-io/conan#648]
Gzip is not really efficient for compressing large amounts of binary data. Also, LZMA has special filters improving compression ratio for executable code. You could use 7z or tar.xz for transmission of packages

**Comment by memsharded** [comment]
Thanks for telling. I think this might be a duplicate to: https://github.com/conan-io/conan/issues/52

We never worked on this feature because a preliminary attempt showed that this will be very complicated from the operations point of view: creating the conan application for PyPi, creating the installers that bundle it for different systems (Win installer, deb, archlinux, OSX homebrew...), as the mainstream python version is still Python2.7, that do not include such functionality.

However, this would be a very nice to have functionality, and very decoupled, not necessary to know about the very internals of conan, if someone knows how to do it and is able to do the work, it will be very welcome.

**Comment by annulen** [comment]
This is not a duplicate of #52 as that issue is talking about use of 7z in conanfile code, though it seems to be a necessary prerequisite.

**Comment by memsharded** [comment]
Yes, you are right. Basically the underlying problem is the same, once provided support for LZMA in conan, such issue #52 would be trivial.

**Comment by tru** [comment]
@memsharded would you be open to having some kind of opt-in to use lzma for packages instead of gzip? lzma makes a LOT of sense for the package storage and transfer, and in our case we can guarantee that all clients are using python3. Would a environment or conan.conf option for uploading .xz packages be something you would accept if we created such a patch?

**Comment by lasote** [comment]
The problem is maintaining the installers and pypi package, the distinction between py2 and py3 to manage will be a pain. Also managing different compressed files on the server side could be very hard to control if the users start to mix them in the public repositories. It's possible but it looks like a nightmare for us.

**Comment by tru** [comment]
I see that - I am not really that worried about the public repos (selfish I know), but in our closed ecosystem this would still be a huge win for us.

**Comment by memsharded** [comment]
I wouldn't oppose to have a branch with it, see how it looks like, and think if there is some way to guarantee 100% that it won't be possible to use it in public repos.

**Comment by lasote** [comment]
I think probably we should probably wait for the definitive dropping of python 2 support. But maybe we should start implementing in the next Conan release a kindly error when a `conan_package.7z` is found to be ready to the future when (probably Conan 2.x) starts to upload 7z packages to public repositories. Then past versions from 1.6 could raise nicely saying: "The package xxx uses 7z compression and your conan version doesn't support it, please upgrade to the latest".
@memsharded @danimtb @markgalpin WDYT?

**Comment by memsharded** [comment]
Yes, I agree, we could start preparing for the future.

**Comment by lasote** [comment]
I'm creating a new issue to my suggestion of being prepared for the future.

**Comment by memsharded** [comment]
We might want to use zstd compression, it is 6 times faster to decompress: https://raw.githubusercontent.com/facebook/zstd/dev/doc/images/DSpeed3.png

Thanks @tru and @sztomi

**Comment by walterschneider** [comment]
Having lzma or zstd would be great!

We have large binary packages where the package-size would decrease significant (gzip: 345M, xz: 218M, zstd: 247M).

Currently we have recursively packed archives (an xz-archive inside the conan_package.tgz) which requires an additional unpack operation after installing the conan package.

**Comment by memsharded** [comment]
@SSE4 could you please put in your backlog do a small profiling experiment of the different compressions? Not urgent, but something we would like to close for Conan 2.0

- Use the binaries (headers, libs) produced by Conan for some popular big packages. Lets use 2-3 like Boost, OpenCV, CGAL.
- Measure the compression time, size and decompression time.
- Use standard (not the Conan one that does other processing) tgz, and the other best alternatives (xz/lzma/zstd) that you think would be real competing alternatives.

@walterschneider our most important and driving metric will be decompression time. Transfers are not the bottleneck, and there are mechanisms to optimize this, like the download cache, but decompression is really the biggest bottleneck in Conan

**Comment by SSE4** [comment]
@memsharded here are some preliminary results of benchmarking gzip/bzip2/lzma/lzo/lz4/zstd on boost/opencv/cgal:

![boost compression time](https://user-images.githubusercontent.com/870236/104338924-a084a600-5529-11eb-9151-9c7dbb5d974f.jpg)
![boost decompression time](https://user-images.githubusercontent.com/870236/104338929-a37f9680-5529-11eb-8490-4f38142ff93c.jpg)
![boost ratio](https://user-images.githubusercontent.com/870236/104338947-a67a8700-5529-11eb-8ec4-40562eb7ba7c.jpg)
![opencv compression time](https://user-images.githubusercontent.com/870236/104338963-ab3f3b00-5529-11eb-8fc3-c43510858e88.jpg)
![opencv decompression time](https://user-images.githubusercontent.com/870236/104338969-ad08fe80-5529-11eb-97d0-a19baade59c3.jpg)
![opencv ratio](https://user-images.githubusercontent.com/870236/104338977-ae3a2b80-5529-11eb-9c6a-733cd85ad6c4.jpg)
![cgal compression time](https://user-images.githubusercontent.com/870236/104339006-b4c8a300-5529-11eb-80ba-83794f6ed621.jpg)
![cgal decompression time](https://user-images.githubusercontent.com/870236/104339013-b6926680-5529-11eb-9417-644273472530.jpg)
![cgal ratio](https://user-images.githubusercontent.com/870236/104339029-ba25ed80-5529-11eb-8972-c48f48e26384.jpg)
NOTE: this is with default compression levels

**Comment by memsharded** [comment]
Good work @SSE4 

My conclusions so far:

- Bzip2 and Lzma are not competitive. They can have great ratios, but are really slow in compressing and decompressing, so not an option.
- Zstd seems a slightly better alternative than gzip in terms of performance. But it is not Python builtin, and PyPI packages like https://pypi.org/project/zstd/ are a pure binding, and that is really problematic for distributing Conan. I cannot see very mainstream zstd implementations that would be very reliable and easy to distribute with Conan
- lzo, lz4 are a bit irregular, sometimes better, sometimes worse than gzip. The ratios seems to be worse, the decompression speed would be worse for boost, better for others. It would be also quite difficult to embed and distribute this with Conan in all platforms.

At the light of these results, my proposal is to keep using tgz format for conan server storage and transfers.

**Comment by tru** [comment]
> Zstd seems a slightly better alternative than gzip in terms of performance. But it is not Python builtin, and PyPI packages like https://pypi.org/project/zstd/ are a pure binding, and that is really problematic for distributing Conan. I cannot see very mainstream zstd implementations that would be very reliable and easy to distribute with Conan

This is true but I think for some users this is very easy to control (like for us).

I think considering that there should be an option to allow users to select the compression format for huge boosts in decompression and storage sizes.

If this can be encoded in the protocol somehow for conan 2 I think that would solve most people's problem. The policy for CCI should still probably be tgz - but you have a lot of users with private installations and that are happy to set something else there for the upsides.

**Comment by tru** [comment]
One thing I didn't see in @SSE4 graphs was the compression setting for zstd? we see way better compression in our local tests - with a bit longer compression time but still the same decompression time. This is probably something that should be tweakable.

**Comment by SSE4** [comment]
@tru all tests were run with default compression levels (it means level 3 for zstd, in particular).
I may also run tests for all compression levels during the night, but it will take way longer.

**Comment by SSE4** [comment]
@memsharded what do you think if we leave gzip a default, but add some options for users who want more control:
```
[compression]
compressor = lzma  # use lzma to compress packages
decompressor = zstd # use zstd to decompress packages
compressor.lzma.level = 9 # use custom compression level for lzma
decompressor.zstd.location = /usr/bin/zstd. # use custom executable for zstd compressor
```

**Comment by memsharded** [comment]
@SSE4 

Knowing the location of the ``zstd`` in the system is still not enough, and would require a very big development effort. If anything, the approach would be more in the line of users needing to provide some ``compress(files, output_filename, symlinks, ...)`` and ``decompress(file, output_folder, ....)``, kind of pluggable/plugins behavior, that implement the compression and decompression they want, under their own responsibility. Because there are many details there, like how symlinks are put and extracted, permissions and other metadata, timestamps... (at the moment the .tgz conan manages clears all timestamps). 

Even implementing this plugin capability is not simple at all, it will require a relevant effort, providing the functionality as built-in as configurable is not something we have resources to do.

**Comment by SSE4** [comment]
@memsharded the idea of tar is that it decouples compression from archiving, so conan can create tar file and use an external command to compress it (zstd, lzo, lz4 or whatever user wants). the same for decompression.
the archiving part handles symlinks, hardlinks, permission, timestamps, etc, and it's always the same code. the only thing changes is external command to run custom compressor or decompressor.
compressor/decompressor only operates on tar files, so it's easy for conan to be compression format agnostic.

**Comment by memsharded** [comment]
Not completely true, for example the timestamps thing is managed in the gz layer, not the tar level (``gzopen_without_timestamps()``)

And then we are using the builtin tarextract python functionality that handles the decompression at once, so we don't need to decompress first, then untar. What I am saying is that this is a very big effort, it is very far from "it's easy for conan to be compression format agnostic". It will require a relevant investment in changing the way things are compressed and decompressed, it will require a new way to define such plugins, to document everything and the interfaces and then to do the maintenance and support that will come with it.

**Comment by tru** [comment]
> @tru all tests were run with default compression levels (it means level 3 for zstd, in particular).

We found that level 9 was good for us.

**Comment by SSE4** [comment]
![boost compression time](https://user-images.githubusercontent.com/870236/104453212-a8e6ea80-55d6-11eb-9b6d-416a11cfed64.jpg)
![boost decompression time](https://user-images.githubusercontent.com/870236/104453222-ab494480-55d6-11eb-8e80-2563e42a5bde.jpg)
![boost ratio](https://user-images.githubusercontent.com/870236/104453226-ac7a7180-55d6-11eb-9fd2-96346a82f03c.jpg)
![cgal compression time](https://user-images.githubusercontent.com/870236/104453230-aedccb80-55d6-11eb-9e4f-2015959f0dec.jpg)
![cgal decompression time](https://user-images.githubusercontent.com/870236/104453234-b00df880-55d6-11eb-81a7-658211922267.jpg)
![cgal ratio](https://user-images.githubusercontent.com/870236/104453240-b1d7bc00-55d6-11eb-8154-3970ed8b9574.jpg)
![opencv compression time](https://user-images.githubusercontent.com/870236/104453244-b2705280-55d6-11eb-8c23-c1bfd1c8d70d.jpg)
![opencv decompression time](https://user-images.githubusercontent.com/870236/104453248-b43a1600-55d6-11eb-9961-a91c3304a9c5.jpg)
![opencv ratio](https://user-images.githubusercontent.com/870236/104453252-b56b4300-55d6-11eb-878f-5f55cc822f1b.jpg)

**Comment by walterschneider** [comment]
I've made some tests with opencv/4.5.1:0e3623c19c9324197de1788dd20f27a1c0725ece:
- downloaded and extracted conan_package.tgz to conan_package.tar
- evaluated with gzip, lzms and zstd (each with the command-line version)

![openvc_sizte](https://user-images.githubusercontent.com/33923060/104473941-cd829880-55bd-11eb-8b9a-0bf520148b1b.png)
![openvc_compression](https://user-images.githubusercontent.com/33923060/104473966-d2dfe300-55bd-11eb-8db4-9fdde15ee1c6.png)
![openvc_dcompression](https://user-images.githubusercontent.com/33923060/104473980-d5dad380-55bd-11eb-96cd-6e85be133490.png)

**Comment by walterschneider** [comment]
i see... @memsharded : thank you for your explanations regarding gzopen_without_timestamps, etc..

**Comment by memsharded** [comment]
Thanks very much all of you for the feedback and the data.

I have been thoroughly consider our options, and at the moment we are not considering this feature:

Yes, there are some gains in compression ratio and decompression speed, for some cases (for zstd). Still not that huge, in many cases it is similar.

The complexity of making it the default approach (making it multi-platform, as this is not Python built-in, nor has a relatively stable native Python package, not a binding, that provides it) makes this option impossible.

Developing a new system of plugins, with the ability that users can connect their own compression and decompression routines is a lot of work, that will introduce fragility, more documentation, maintenance, support...

We really need to focus our development efforts in having a very solid Conan 2.0, and there are millions of critical, very required to-do things: from better integration with the build systems, to a more powerful graph model. Unless something it is really not working and needs to be fixed, it would need to wait for later. It could even be Conan 2.1, 2.2.... but not Conan 2.0, that is a different beast. Only managing a as smooth as possible transition from 1.X will be a huge effort.

**Comment by memsharded** [comment]
Closing, at the moment this will not be considered.

**Comment by DS-Serafin** [comment]
Just wanted to bum this / add some context. The current compression seems very inefficient on certain packages. Compressing a build of grpc oder google-cloud-cpp takes multiple minutes on my mac. This is very anoying and in the range of brocken in my opinion.

**Comment by memsharded** [comment]
Hi @DS-Serafin 

But have you read the above rationale, that made us close this issue? We know this can be slow in some cases, but the thing is that a package should be compressed once, but they will be decompressed many many times. So the compression algorithm has to be chosen based on the decompression speed, not on the compression speed. Using LZMA optimizes for size, not compression/decompression speed, so this is certainly not a solution, using gzip, as Conan is already doing, offers the best trade-offs.

**Comment by sztomi** [comment]
@memsharded I'd argue that zstd offers the best tradeoffs, as it was shown in this thread. It is several times faster than zlib/gzip for decompression and offers a better compression/speed ratio consistently.

**Comment by memsharded** [comment]
> It is several times faster than zlib/gzip for decompression and offers a better compression/speed ratio consistently.

Several times faster in decompression? This is not what the graphs above are showing. They consistently show decompression times in the same order of magnitude, actually really close. It seems like it would be slightly faster, maybe 5-10%, but not several times faster.

And the complexities for integrating such a compression algorithm in the Conan app would be huge, introducing a high fragility, tons of issues with non mainstream platforms, etc. Definitely not worth it, we still strongly believe that gzip is the best trade-off

**Comment by DS-Serafin** [comment]
@memsharded I understand the rational. And it makes sense, but I think the current settings are a bit to extreme. I mayself package a lot and upload a lot of 3rd party libs to our internal artifactory. So while yes decrompession is more important than compression, I think we should have an algorithm / compression settings which will not completly deteroate when presented with something other than a normal text file.

**Comment by sztomi** [comment]
> Several times faster in decompression? 

Yes. gzip (i.e. zlib/DEFLATE) decompresses at about 400 MB/s, whereas zstandard performs 1700-2300 MB/s on the same machine. And it's possible that for low entropy data the gap is even bigger because zlib slows down on such input. 

https://github.com/facebook/zstd#benchmarks

**Comment by memsharded** [comment]
> Yes. gzip (i.e. zlib/DEFLATE) decompresses at about 400 MB/s, whereas zstandard performs 1700-2300 MB/s on the same machine. And it's possible that for low entropy data the gap is even bigger because zlib slows down on such input.

But this is not real benchmarking for unzipping typical libraries and artifacts from C/C++, where many other factors need to be taken into account. That is the "pure" throughput of the algorithm, but the graphs above show that in practice their application for decompressing Conan packages, there is not such a big difference.

**Comment by sztomi** [comment]
We are using it to decompress a bundle of conan deps and it is very noticably faster than everything else (even though we increase the compression level from the default). I think this is due to zstd having better I/O management.

**Comment by grossag** [comment]
I am wondering if zstd could be an additional, non-default, compression format used. For example, if I do a `conan upload --binary-compression-format zstd`, the script could expect me to have `tar` and `zstd` in my `PATH`. The default case could still be a pure-Python implementation that works out of the box, but `zstd` support could require a bit more configuration for a nice performance gain, especially in decompression.

In our case, we are using conan to slice up the Visual Studio build tools to only stage the portion that we use. This is 7GB uncompressed and results in a 1.8GB .tar.gz file. The compression takes 25-30 minutes. I haven't tested decompression yet, but I expect it to be quite slow.

The zstd binary doesn't have support for recursively compressing binaries, so that makes things more difficult. In practice, you have to use `tar` and request zstd compression, which causes it to then call into `zstd` as a subprocess.

**Comment by memsharded** [comment]
Thanks for your feedback @grossag 

> I am wondering if zstd could be an additional, non-default, compression format used. For example, if I do a conan upload --binary-compression-format zstd, the script could expect me to have tar and zstd in my PATH. The default case could still be a pure-Python implementation that works out of the box, but zstd support could require a bit more configuration for a nice performance gain, especially in decompression.

This brings a lot of extra corner cases. What if you have in the cache already a gz, but then enable the opt-in zstd, now you have 2 files? Will the 2 files be uploaded to the server? What happens if the server contains a zstd but the client doesn't support it? I see users requesting the feature to be able to host both formats, then the client selecting dynamically which one to download (which would be a challenging feature too)

> The zstd binary doesn't have support for recursively compressing binaries, so that makes things more difficult. In practice, you have to use tar and request zstd compression, which causes it to then call into zstd as a subprocess.

This kind of complexity is what is mostly unmanageable by the team, at least at this moment. This kind of feature generate lots of debugging, support, maintenance, documentation in the short, mid and long term.

I am sorry to be the one that brings the bad news, but we have considered this not only once, but rather a few times, and always reached the same conclusion. It would be a very large effort and problematic investment for some performance gains, not even functional changes. I know you are fine with that extra setup from your side, but once this door is open, many users will expect that Conan manages this for them automatically, and will be upset if we don't. We have to release self-contained bundled conan executables, because for many users installing the ``pip install conan`` is already a no-go.

There is also the viral factor, there are already quite a few Conan users that are generating packages and sharing them with another organization, via some shared repos. If the first org uses zstd compression, but the second org cannot easily install themselves the zstd support in their developers machines, Conan will be blamed for it, and the pressure will be on us to automate the distribution of ``tar`` and ``zstd`` executables in all platforms Conan runs to enable and support this feature. A nightmare we really don't want to get there.


> n our case, we are using conan to slice up the Visual Studio build tools to only stage the portion that we use. This is 7GB uncompressed and results in a 1.8GB .tar.gz file. The compression takes 25-30 minutes. I haven't tested decompression yet, but I expect it to be quite slow.

Please let us know about the comparative results with gz and zstd, this would be insteresting (including the tar time, of course, to be a fair comparison).

We are open to explore other approaches to minimize the impact. For example if developers install this ``tool_requires`` just once every few months, the fact that it takes 25 or 40 minutes is not a huge difference. Maybe the ``--lru`` cache cleaning strategy (planned for next releases) can help developers to clean their cache periodically, not needing to install VS-build tools every day. Which in any case, should be the goal, if installing the 7Gb build tools at every new ``conan install`` command, then, it would be a problem no matter the compression algorithm.

**Comment by grossag** [comment]
Thanks for the detailed answer. I understand where you are coming from in wanting to minimize support issues.

On our side, we have ephemeral build machines that need to sync packages every time. We want to minimize the decompression time as much as possible because our build pipeline will be doing a `conan install` every time. I agree that developers wouldn't be doing this all that often, as we only rev the compiler every few months.

A few months ago I had investigating fixing some issues with urllib3's usage of python-zstandard so I have some familiarity with those APIs. It is possible to do a pure Python integration with python-zstandard that wouldn't have the tooling issues that using `tar` and `zstd` directly would.

I have .tar.zst decompression working with a pure Python approach and it's actually faster than using `tar` directly. This is likely because python-zstandard uses a C extension when run in CPython, so everything is in-proc, while `tar` seems to pipe to/from `zstd`. I'm looking into compression now.

I'll post some performance results and a link to the code on my fork in the next week or two in case this is something you'd be interested in me pursuing as a pull request.

**Comment by grossag** [comment]
Here is what my initial change looks like: https://github.com/grossag/conan/commit/278ea0269dfc8d3d76eee13b7bf0eec3a704f7d8

It would be possible to support multiple compression formats in the setting if you think teams would want to do both gzip and zstd compression to handle users of older version of conan. I didn't put that in there yet, because I don't think people would want to use the extra storage. I also am new at Conan so I don't understand the package manifest file formats and how all the hashes are calculated.

Let me know if you want me to create a pull request for it so we can have more discussion there.

Initial results, all done at the default compression levels:
1. Compression is 17-20x faster and results in binaries that are about 20% smaller. Example from today: 7GB compiler toolset with gzip takes 2100 seconds to produce a 1.84GB .tgz file, and with zstd takes 119 seconds to produce a 1.5GB .tar.zst file
2. Decompression performance improvement varies. I don't have a great test harness. So far my tests show that decompression is 10-30% faster. Example from today: decompressing the .tgz file from above took 78 seconds seconds, and with zstd it took 45 seconds seconds.

**Comment by memsharded** [comment]
Thanks for the work and sharing it.

I am fine with opening a PR and start talking about this, but please note:

- This will most likely fail in CI, because there might be no provision for zstd in the servers there. The team is quite swamped, it might take a while until we can do that.
- The concerns about the infrastructure, and depending on zstandard which is 0.X unstable version, with a clear warning that they will break in the future, is problematic. Support for OSX is not defined in their page, which is also problematic. 

The results you shared are relevant. Compression speed is not that critical, at the end of the day, it is only necessary to compress one, and it is decompression speed what is the critical thing. Saving a 10% of decompression time is definitely not worth the investment and extra complexity. Average 30% over a wide range of real world packages, that might be more interesting.

Overall, I don't want to give false hopes that this will be merged, the risks are still too high for the value. I know it feels bad putting effort into something that finally doesn't get merged, please take into account that while it is good to have the PR to discuss things and learn, there are no guarantees that this will be moved forward. Several years back it would have had more possibilities, but now we have too many scars from supporting and evolving Conan for that long, that risky things are discarded as much as possible, specially if they don't benefit to the majority of Conan users.

**Comment by grossag** [comment]
I'll create the PR, thanks. FWIW they do support MacOS. We have been using it internally in our builds as part of remote caching of artifacts for Windows, Mac, and Linux in two ways:

1. urllib3 2.0 supports zstd decompression for GET requests using python-zstandard
2. We use python-zstandard directly for compressing PUT requests

I do still think there is value here in that as long as the package producer chooses zstd (or the default for producers is switched some day in the future), the consumers will get it automatically.

**Comment by Ext3h** [comment]
> Average 30% over a wide range of real world packages, that might be more interesting.

It's actually significantly more than that. More like another flat out factor ~2 speedup. The `zstd` implementation is hit dis-proportionally harder by the flat cost on the `tarfile.‎TarFile` side. The interleaved design of `tarfile.‎TarFile` turns an embarrassingly parallel problem into a serial one.

**Comment by Ext3h** [comment]
> depending on zstandard which is 0.X unstable version, with a clear warning that they will break in the future, is problematic. Support for OSX is not defined in their page, which is also problematic.

On the other hand, the relevant interfaces `zstandard.ZstdCompressionWriter` and `ZstdDecompressionReader` are mature, as far as I can tell after checking both the implementation and the bugtracker. No breaking changes expected, as these interfaces are implementing `io.RawIOBase`.

The warnings relate to the interfaces `ZstdCompressionObj` and `ZstdDecompressionObj` which got details about the handling of frames in zstd wrong on the interface definition level.

**Comment by memsharded** [comment]
I am resuming work on this, now that Python 3.14 has built-in zstd compression.

The current PR, derived from this one is in: https://github.com/conan-io/conan/pull/19337

We are planning to try to include it in Conan 2.25, in January 2026. It would be very useful to have testing and feedback, so if you want to experiment running from source from my branch in https://github.com/conan-io/conan/pull/19337, that would be very useful. Many thanks!

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/19337 has been merged, implementing both zstd (for Python>=3.14) and xz compressions built-in. It will be released as experimental in next Conan 2.25.
